src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -934,6 +1116,205 @@ - (void)readAttributesWithEndpointID:(NSNumber * _Nullable)endpointID
+            reportHandler(nil, [MTRError errorForCHIPErrorCode:CHIP_ERROR_INVALID_ARGUMENT]);
+        });
+        return;
+    }
+
+    if (self.isPASEDevice) {
+        // We don't support subscriptions over PASE.
+        dispatch_async(queue, ^{
+            reportHandler(nil, [MTRError errorForCHIPErrorCode:CHIP_ERROR_INCORRECT_STATE]);
+        });
+        return;
+    }
+
+    // Copy params before going async.
+    attributePaths = (attributePaths == nil) ? nil : [attributePaths copy];
+    eventPaths = (eventPaths == nil) ? nil : [eventPaths copy];
+    params = (params == nil) ? nil : [params copy];
+
+    [self.deviceController
+        getSessionForNode:self.nodeID
+               completion:^(ExchangeManager * _Nullable exchangeManager, const Optional<SessionHandle> & session,
+                   NSError * _Nullable error) {
+                   if (error != nil) {
+                       if (reportHandler) {
+                           dispatch_async(queue, ^{
+                               reportHandler(nil, error);
+                           });
+                       }
+                       return;
+                   }
+
+                   auto onAttributeReportCb
+                       = [queue, reportHandler](const app::ConcreteClusterPath & clusterPath, const AttributeId aAttributeId,
+                             const MTRDataValueDictionaryDecodableType & data) {
+                             id valueObject = data.GetDecodedObject();
+                             app::ConcreteAttributePath pathCopy(clusterPath.mEndpointId, clusterPath.mClusterId, aAttributeId);
+                             dispatch_async(queue, ^{
+                                 reportHandler(@[ @ {
+                                     MTRAttributePathKey : [[MTRAttributePath alloc] initWithPath:pathCopy],
+                                     MTRDataKey : valueObject
+                                 } ],
+                                     nil);
+                             });
+                         };
+
+                   auto onEventReportCb = [queue, reportHandler](const app::ConcreteClusterPath & clusterPath,
+                                              const EventId aEventId, const MTRDataValueDictionaryDecodableType & data) {
+                       id valueObject = data.GetDecodedObject();
+                       app::ConcreteEventPath pathCopy(clusterPath.mEndpointId, clusterPath.mClusterId, aEventId);
+                       dispatch_async(queue, ^{
+                           reportHandler(
+                               @[ @ { MTREventPathKey : [[MTREventPath alloc] initWithPath:pathCopy], MTRDataKey : valueObject } ],
+                               nil);
+                       });
+                   };
+
+                   auto establishedOrFailed = chip::Platform::MakeShared<BOOL>(NO);
+                   auto onFailureCb = [establishedOrFailed, queue, subscriptionEstablished, reportHandler](
```

Comment:
I know you copied this from the existing subscribeAttributes... code, but if `clusterPath` is not null here, this logic is wrong...  Please add a TODO comment and I will fix that later.

Reply:
I added.